Added API Guidance (#45)
DFE-Digital/technical-guidance: guides/api-guidance.md

---
category: API's
expires: 2021-06-23
---

# API Guidance: Developing a DfE API

> API's are an increasingly important part of DFE services. We are supporting an API first approach to help make our services highly operable and automated. API's, however, should start with a need or a problem to solve. They can be used in many different ways. They can be used to integrate systems and services, collect or share data, consume and publish events as well as consume other API's.

Reviewer: ```suggestion
Application programmable interfaces or APIs are an increasingly important part of DfE services. We are supporting an API first approach to help make our services highly operable and automated. APIs, however, should start with a need or a problem to solve. They can be used in many different ways. They can be used to integrate systems and services, expose functionality, collect or share data, consume and publish events as well as consume other APIs.
```

There are a number of different ways to create an API, this guide does not give step by step guidance but should be used to help the process and ensure your API is written to DfE/GDS guidlines and standards.

## User Needs

Understand who your users are and what their needs are, consider their needs in context of the end to end service. Ensure you fully understand the problem space. For example your API might be intended to be consumed by an educational software supplier who are certainly a key user, but are they the only users you should be considering when designing an API? Maybe not, ask yourself who are the end users? It could be a teacher, a pupil, a school administrator for example. Is your API designed in a way that considers their needs as well as the software vendor?

Fully understanding users and their needs is the most important ingredient to ensuring your API is fit for purpose. Work with you user not just to define the API design but also ask consumers what is sufficient to test your API. Ensure your API is testable and ceate a sandbox including aspects such as authorisation if applicable.

> If your API is intended to be used by DfE suppliers and providers then make sure you establish working relationships with them at the begining of your API journey and ensure your service is fully aware of their responsibilities of maintaing these relationships throughout the life of the API. 

Reviewer: ```suggestion
If your API is intended to be used by DfE suppliers and providers then make sure you establish working relationships with them at the beginning of your API journey and ensure your service is fully aware of their responsibilities of maintaining these relationships throughout the life of the API. 
```

[GOV.UK API Design Guidance](https://www.gov.uk/government/collections/api-design-guidance)

Design around your users and your data. The illustration below shows a simple design pattern. Understanding data flow, security and master / authoritive sources is a key part of the design process. For example if your API exposes data to consumers then it ensures you expose the right data from the right source.

![Simple API design pattern for DfE services consuming / publishing data](technical-guidance/images/api-service-pattern.png)

## Naming an API

API names are important, especially those API's that are external as they are representing the DfE and need to be reflective of the API's purpose. Your API name is an opportunity to convey a lot of information in a concise manner. It helps to set the context for what to expect within the API. The name must also adhere to GDS standards in order to be granted an api.gov.uk sub domain name.

First, follow the GDS guidance here [When naming an API](https://www.gov.uk/guidance/gds-api-technical-and-data-standards#when-naming-and-hosting-your-api)

### DfE API naming convention

Use the below convention for external facing DfE API's:

> <https://education-{api-name}.api.gov.uk/{resource>}

Reviewer: ```suggestion
`<https://education-{api-name}.api.gov.uk/{resource>}`
```

Example:

> <https://education-school-attendance.api.gov.uk/submission>

Reviewer: ```suggestion
`<https://education-school-attendance.api.gov.uk/submission>`
```

* [DfE developer hub](https://dfe-developerhub.education.gov.uk) A new service that provides a GOV.UK branded public facing front door to DfE's API's. A place to publish your DfE API's, document them, provide information on authentication and access requests. It also provides a sandbox to allow potential users to test your API.

* [DfE- Digital GitHub Repo](https://github.com/DFE-Digital)
    Git Hub repository containing some of DfE's API source and documentation.

* [GOV.UK API catalogue](https://www.api.gov.uk/#uk-government-apis) A new ALPHA service providing a cross Governmen view of available API's (If you publish here please just list your API and sign post back to the [DfE developer hub](https://dfe-developerhub.education.gov.uk)) for documentation, testing etc.

The DfE has a number of disperate API catalogues/repo's. We have recognised this and are working to rationalise these using the [DfE developer hub](https://dfe-developerhub.education.gov.uk) to make it easier and provide a consistent way for potential users and publishers of API's to get connected.

## API Specification and documentation

> Use [OAS](https://swagger.io/specification/) (Open API Standard v3) to specify your RESTFUL API. It defines a standard, language-agnostic interface to RESTful APIs which allows both humans and computers to discover and understand the capabilities of the service without access to source code, documentation, or through network traffic inspection. When properly defined, a consumer can understand and interact with the remote service with a minimal amount of implementation logic.

Reviewer: ```suggestion
Use [OAS](https://swagger.io/specification/) (Open API Standard v3) to specify your RESTful API. It defines a standard, language-agnostic interface to RESTful APIs which allows both humans and computers to discover and understand the capabilities of the service without access to source code, documentation, or through network traffic inspection. When properly defined, a consumer can understand and interact with the remote service with a minimal amount of implementation logic.
```

> An OpenAPI definition can then be used by documentation generation tools to display the API, code generation tools to generate servers and clients in various programming languages, testing tools, and many other use cases.Publish your documentation with your API and use [DfE developer hub](https://dfe-developerhub.education.gov.uk).

Reviewer: ```suggestion
An OpenAPI definition can then be used by documentation generation tools to display the API, code generation tools to generate servers and clients in various programming languages, testing tools, and many other use cases. Publish your documentation with your API and use [DfE developer hub](https://dfe-developerhub.education.gov.uk).
```

## API Security

Make sure that your API is secure. GDS provides specific guidlines [here](https://www.gov.uk/guidance/gds-api-technical-and-data-standards#when-to-authenticate-your-api). As well as this general guide, we have provided specific guidance here on securing your DfE API [here](https://dfe-developerhub.education.gov.uk/documentation/authorisation)

## Understand your data

> If your API is providing access to DfE data then ensure it is done so securely and it is data that can be shared (you should contact the DfE CIRO office to check). Never expose sensitive data unless you have implemented the necessary authentication and that the data owner is fully aware and has approved its publication.

Reviewer: ```suggestion
If your API is providing access to DfE data then ensure it is done so securely and it is data that can be shared (you should contact the DfE SIRO office to check). Never expose sensitive data unless you have implemented the necessary authentication and that the data owner is fully aware and has approved its publication.
```

Min/max values, defaults when the value isn’t provided, and other related details.

## Return Codes (HTTP)

Use standard http return codes as per the list maintained by [W3C](https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html). 

All API response status codes are separated into five classes (or categories) in a global standard. The first digit of the status code defines the class of response. The last two digits do not have any class or categorization role.

* 1xx (Informational): The request is received and continues to be processed
* 2xx (Successful): The request is successfully received, understood, and accepted
* 3xx (Redirection): Further action needs to be taken to complete the request
* 4xx (Client Error): The request contains the wrong syntax or cannot be fulfilled
* 5xx (Server Error): The server fails to fulfill an apparently valid request

## Testing

To successfully test an API, you need to fully understand its purpose and workflow. Leverage toolsets and automate where possible. Test your API’s compliance, performance and scalability and create positive and negative tests.

## Tools

> IF you are an experienced API developer you will already have a preffered toolset. For those starting out creating API's here is a list of useful tools to check out..

Reviewer 2: ```suggestion
If you are an experienced API developer you will already have a preferred toolset. For those starting out creating API's here is a list of useful tools to check out.
```

> IF you are an experienced API developer you will already have a preffered toolset. For those starting out creating API's here is a list of useful tools to check out..

Reviewer: ```suggestion
IF you are an experienced API developer you will already have a preferred toolset. For those starting out creating API's here is a list of useful tools to check out..
```
